test/lib/jdk/test/lib/hotspot/ir_framework/TestFramework.java

```diff
@@ -0,0 +1,1044 @@
+        }
+
+        flagList.addAll(Arrays.asList(getDefaultFlags()));
+
+        if (shouldVerifyIR) {
+            String flags = socket.getOutput();
+            if (VERBOSE) {
+                System.out.println("Read sent data from flag VM from socket:");
+                System.out.println(flags);
+            }
+            String patternString = "(?<=" + TestFramework.TEST_VM_FLAGS_START + "\\R)" + "(.*DShouldDoIRVerification=(true|false).*)\\R"
+                                   + "(?=" + IREncodingPrinter.END + ")";
+            Pattern pattern = Pattern.compile(patternString);
+            Matcher matcher = pattern.matcher(flags);
+            check(matcher.find(), "Invalid flag encoding emitted by flag VM");
+            // Maybe we run with flags that make IR verification impossible
+            shouldVerifyIR = Boolean.parseBoolean(matcher.group(2));
+            flagList.addAll(Arrays.asList(matcher.group(1).split(TEST_VM_FLAGS_DELIMITER)));
+        }
+        return flagList;
+    }
+
+    private void checkTestVMExitCode(JVMOutput vmOutput) {
+        final int exitCode = vmOutput.getExitCode();
+        if (EXCLUDE_RANDOM || REPORT_STDOUT || (VERBOSE && exitCode == 0)) {
+            System.out.println("--- OUTPUT TestFramework test VM ---");
+            System.out.println(vmOutput.getOutput());
+        }
+
+        if (exitCode != 0) {
+            throwTestVMException(vmOutput);
+        }
+    }
+
+    private void throwTestVMException(JVMOutput vmOutput) {
+        String stdErr = vmOutput.getStderr();
+        if (stdErr.contains("TestFormat.reportIfAnyFailures")) {
+            Pattern pattern = Pattern.compile("Violations \\(\\d+\\)[\\s\\S]*(?=/============/)");
+            Matcher matcher = pattern.matcher(stdErr);
+            TestFramework.check(matcher.find(), "Must find violation matches");
+            throw new TestFormatException(System.lineSeparator() + System.lineSeparator() + matcher.group());
+        } else if (stdErr.contains("NoTestsRunException")) {
+            shouldVerifyIR = false;
+            throw new NoTestsRunException(">>> No tests run due to empty set specified with -DTest and/or -DExclude. " +
+                                          "Make sure to define a set of at least one @Test method");
+        } else {
+            throw new TestVMException(vmOutput.getExceptionInfo(scenario != null));
+        }
+    }
+
+    static void check(boolean test, String failureMessage) {
+        if (!test) {
+            throw new TestFrameworkException(failureMessage);
+        }
+    }
+}
+
+/**
+ * Class to encapsulate information about the test VM output, the run process and the scenario.
+ */
+class JVMOutput {
+    private final Scenario scenario;
+    private final OutputAnalyzer oa;
+    private final ProcessBuilder process;
+    private final String hotspotPidFileName;
+
+    JVMOutput(OutputAnalyzer oa, Scenario scenario, ProcessBuilder process) {
```

Comment:
instead of passing a ProcessBuilder, you can either pass a command-line or better just past prepared flags, and make it JVMOutput's (you will need a better name) responsibility to execute the process, create OutputAnalyzer, etc. 

```suggestion
    JVMOutput(List<String> cmds, Scenario scenario) {
    var pb = ProcessTools.createJavaProcessBuilder(cmds);
    try {
            // Calls 'main' of TestFrameworkExecution to run all specified tests with commands 'cmds'.
            // Use executeProcess instead of executeTestJvm as we have already added the JTreg VM and
            // Java options in prepareTestVMFlags().
            this.oa = ProcessTools.executeProcess(process);
        } catch (Exception e) {
            throw new TestFrameworkException("Error while executing Test VM", e);
        }
        this.cmd = pb.command();
        ...
```

and then runTest will be smth like
```
    private void runTestVM(List<String> additionalFlags) {
          List<String> cmds = prepareTestVMFlags(additionalFlags);
          socket.start();
          JVMOutput output = new JVMOutput(cmds, scenario);
          ...
```

Reply:
I did a refactoring and introduced new classes `FlagVMProcess` and `TestVMProcess`. I also did a renaming of `TestFrameworkPrepareFlags` -> `FlagVM` and `TestFrameworkExecution` -> `TestVM` to make it more clear whats going on.

Reply:
I did a refactoring and introduced new classes `FlagVMProcess` and `TestVMProcess`. I also did a renaming of `TestFrameworkPrepareFlags` -> `FlagVM` and `TestFrameworkExecution` -> `TestVM` to make it more clear whats going on.